runtime/src/iree/testing/benchmark.h

```diff
@@ -12,12 +12,97 @@
 // systems and use some simple tooling while also allowing them to run on
 // the full benchmark library with all its useful reporting and statistics.
 
+#include <math.h>
+
 #include "iree/base/api.h"
 
 #ifdef __cplusplus
 extern "C" {
 #endif  // __cplusplus
 
+//===----------------------------------------------------------------------===//
+// Benchmarking tools
+//===----------------------------------------------------------------------===//
+
+void iree_benchmark_use_ptr(char const volatile*);
+
+#if !defined(IREE_BENCHMARK_HAS_INLINE_ASSEMBLY)
+#if defined(IREE_COMPILER_MSVC) || defined(IREE_PLATFORM_EMSCRIPTEN)
+#define IREE_BENCHMARK_HAS_INLINE_ASSEMBLY 0
+#elif defined(IREE_COMPILER_CLANG) || defined(IREE_COMPILER_GCC)
+#define IREE_BENCHMARK_HAS_INLINE_ASSEMBLY 1
+#else
+#define IREE_BENCHMARK_HAS_INLINE_ASSEMBLY 0
+#endif  // non-asm-targets
+#endif  // !IREE_BENCHMARK_HAS_INLINE_ASSEMBLY
+
+#if IREE_BENCHMARK_HAS_INLINE_ASSEMBLY == 0
+
+#if defined(IREE_COMPILER_MSVC)
+#define iree_benchmark_clobber() _ReadWriteBarrier()
+#else
+#define iree_benchmark_clobber()
+#endif  // IREE_COMPILER_MSVC
+
+#if defined(__cplusplus)
+}  // extern "C"
+template <typename T>
+inline IREE_ATTRIBUTE_ALWAYS_INLINE void iree_optimization_barrier(T&& value) {
+  iree_benchmark_use_ptr(&reinterpret_cast<char const volatile&>(value));
+  iree_benchmark_clobber();
+}
+extern "C" {
+#else
+// TODO: a C-compatible optimization barrier.
+#define iree_optimization_barrier(x)
+#endif  // __cplusplus
+
+#elif defined(IREE_COMPILER_CLANG)
+
+#if defined(__cplusplus)
+}  // extern "C"
+inline IREE_ATTRIBUTE_ALWAYS_INLINE void iree_benchmark_clobber() {
+  asm volatile("" : : : "memory");
+}
+template <class T>
+inline IREE_ATTRIBUTE_ALWAYS_INLINE void iree_optimization_barrier(T&& value) {
+  asm volatile("" : "+r,m"(value) : : "memory");
+}
+extern "C" {
+#else
+// TODO: a C-compatible optimization barrier.
+#define iree_optimization_barrier(x)
+#endif  // __cplusplus
+
+#elif defined(IREE_COMPILER_GCC)
+
+#if defined(__cplusplus)
+}  // extern "C"
+inline IREE_ATTRIBUTE_ALWAYS_INLINE void iree_benchmark_clobber() {
+  asm volatile("" : : : "memory");
+}
+template <class T>
+inline IREE_ATTRIBUTE_ALWAYS_INLINE
+    typename std::enable_if<std::is_trivially_copyable<Tp>::value &&
+                            (sizeof(T) <= sizeof(T*))>::type
```

Comment:
typo? https://github.com/iree-org/iree/actions/runs/9024672443/job/24798990302?pr=17336#step:4:8404
```
In file included from /work/runtime/src/iree/vm/bytecode/module_benchmark.cc:11:
/work/runtime/src/iree/testing/benchmark.h:86:56: error: 'Tp' was not declared in this scope; did you mean 'T'?
   86 |     typename std::enable_if<std::is_trivially_copyable<Tp>::value &&
      |                                                        ^~
      |                                                        T
```

```suggestion
template <class T>
inline IREE_ATTRIBUTE_ALWAYS_INLINE
    typename std::enable_if<std::is_trivially_copyable<T>::value &&
                            (sizeof(T) <= sizeof(T*))>::type
```